File: scripts/data/extract_1990_municipal_winners_from_docx.ts
/**
 * Phase H1.2.2: data:extract1990 generator — Excel → municipality_political_controllers.json.
 * Reads data/source/1990 to 1995 municipalities_BiH.xlsx (Municipality, Party that won 1990 elections).
 * Output: data/source/municipality_political_controllers.json ({ version, mappings }).
 */

import { readFile, writeFile } from 'node:fs/promises';
import { resolve, join } from 'node:path';
import { fileURLToPath } from 'node:url';
import { dirname } from 'node:path';
import { existsSync } from 'node:fs';
import * as XLSX from 'xlsx';



const __dirname = dirname(fileURLToPath(import.meta.url));
const ROOT = resolve(__dirname, '../..');


const DEFAULT_INPUT = 'data/source/1990 to 1995 municipalities_BiH.xlsx';
const OUTPUT_PATH = join(ROOT, 'data/source/municipality_political_controllers.json');
const INDEX_PATH = join(ROOT, 'data/source/municipality_post1995_to_mun1990.json');
const CANONICAL_IDS = ['RBiH', 'RS', 'HRHB'] as const;
type ControllerId = (typeof CANONICAL_IDS)[number];

/** Party name (normalized) → controller. Full names from Excel "Party that won 1990 elections". Ledger 6B.2: SDA/SK-SDP→RBiH, SDS→RS, HDZ BiH→HRHB. */
const PARTY_TO_CONTROLLER: Record<string, ControllerId> = {
  'stranka demokratske akcije': 'RBiH',
  'srpska demokratska stranka': 'RS',
  'hrvatska demokratska zajednica bih': 'HRHB',
  'hrvatska demokratska zajednica': 'HRHB',
  'savez komunista bih - stranka demokratskih promjena': 'RBiH',
  'stranka demokratskih promjena': 'RBiH',
  sda: 'RBiH',
  sds: 'RS',
  'hdz bih': 'HRHB',
  hdz: 'HRHB',
  'sk-sdp': 'RBiH'
};

function normalizeWhitespace(s: string): string {
  return s.replace(/\s+/g, ' ').trim();
}

function normalizeForMatch(s: string): string {
  return normalizeWhitespace(s).toLowerCase();
}

function parseArgs(): { input: string } {
  const args = process.argv.slice(2);
  let input = join(ROOT, DEFAULT_INPUT);
  for (let i = 0; i < args.length; i++) {
    if (args[i] === '--input' && args[i + 1]) {
      const next = args[++i];
      input = next.startsWith('/') || /^[A-Za-z]:/.test(next) ? next : join(ROOT, next);
      break;
    }
  }
  return { input };
}

interface RemapRow {
  post1995_code: string;
  post1995_name: string;
  mun1990_name?: string;
}

interface RemapFile {
  rows?: RemapRow[];
  index_by_post1995_code?: Record<string, string>;
}

function buildNameToCodes(remap: RemapFile): Map<string, string[]> {
  const map = new Map<string, string[]>();
  const rows = remap.rows ?? [];
  for (const row of rows) {
    const code = row.post1995_code;
    if (!code) continue;
    for (const name of [row.post1995_name, row.mun1990_name].filter(Boolean) as string[]) {
      const key = normalizeForMatch(name);
      if (!key) continue;
      const list = map.get(key) ?? [];
      if (!list.includes(code)) list.push(code);
      map.set(key, list);
    }
  }
  for (const list of map.values()) list.sort((a, b) => a.localeCompare(b));
  return map;
}

function partyToController(partyRaw: string): ControllerId | null {
  const key = normalizeForMatch(partyRaw);
  if (!key) return null;
  if (PARTY_TO_CONTROLLER[key]) return PARTY_TO_CONTROLLER[key];
  if (key.includes('demokratske akcije') || key.includes('sda')) return 'RBiH';
  if (key.includes('srpska demokratska') || key.includes('sds')) return 'RS';
  if (key.includes('hrvatska demokratska') || key.includes('hdz')) return 'HRHB';
  if (key.includes('demokratskih promjena') || key.includes('komunista')) return 'RBiH'; // SK-SDP
  return null;
}

async function extractFromExcel(inputPath: string): Promise<Record<string, ControllerId>> {
  const buffer = await readFile(inputPath);
  const workbook = XLSX.read(buffer, { type: 'buffer' });
  const sheetName = workbook.SheetNames[0];
  if (!sheetName) throw new Error('Excel has no sheets');
  const sheet = workbook.Sheets[sheetName];
  const ref = sheet['!ref'];
  if (!ref) throw new Error('Sheet has no range');

  const range = XLSX.utils.decode_range(ref);
  const headerRow = range.s.r;
  const headers: string[] = [];
  for (let c = range.s.c; c <= range.e.c; c++) {
    const cell = sheet[XLSX.utils.encode_cell({ r: headerRow, c })];
    const val = cell && cell.v != null ? normalizeWhitespace(String(cell.v)) : '';
    headers.push(val);
  }

  const muniHeader = headers.find(
    (h) =>
      /municipality/i.test(h) ||
      /op[cć]ina/i.test(h) ||
      /opstina/i.test(h)
  );
  const partyHeader = headers.find(
    (h) =>
      /party/i.test(h) ||
      /stranka/i.test(h) ||
      /winner/i.test(h) ||
      /1990/i.test(h)
  );
  if (!muniHeader || !partyHeader) {
    throw new Error(
      `Required columns not found. Need municipality-like and party-like. Headers: ${headers.join(', ')}`
    );
  }
  const colMuni = headers.indexOf(muniHeader);
  const colParty = headers.indexOf(partyHeader);

  const remapRaw = await readFile(INDEX_PATH, 'utf8');
  const remap = JSON.parse(remapRaw) as RemapFile;
  const nameToCodes = buildNameToCodes(remap);

  const mapping: Record<string, ControllerId> = {};
  const unknownParties = new Set<string>();
  const unmatchedMunicipalities = new Set<string>();

  for (let r = headerRow + 1; r <= range.e.r; r++) {
    const cellMuni = sheet[XLSX.utils.encode_cell({ r, c: colMuni })];
    const cellParty = sheet[XLSX.utils.encode_cell({ r, c: colParty })];
    const muniRaw = cellMuni && cellMuni.v != null ? normalizeWhitespace(String(cellMuni.v)) : '';
    const partyRaw = cellParty && cellParty.v != null ? normalizeWhitespace(String(cellParty.v)) : '';
    if (!muniRaw || !partyRaw) continue;

    const controller = partyToController(partyRaw);
    if (!controller) {
      unknownParties.add(partyRaw);
      continue;
    }

    const key = normalizeForMatch(muniRaw);
    const codes = nameToCodes.get(key);
    if (!codes || codes.length === 0) {
      unmatchedMunicipalities.add(muniRaw);
      continue;
    }
    for (const code of codes) {
      mapping[code] = controller;
    }
  }

  if (unknownParties.size > 0) {
    throw new Error(
      `Unknown party names (cannot map to RBiH/RS/HRHB): ${[...unknownParties].sort().join(', ')}`
    );
  }
  if (Object.keys(mapping).length === 0) {
    throw new Error('Mapping is empty. Check Excel columns and remap index.');
  }
  if (unmatchedMunicipalities.size > 0) {
    process.stderr.write(
      `Warning: Unmatched municipalities (not in remap): ${[...unmatchedMunicipalities].sort().join(', ')}\n`
    );
  }

  return mapping;
}

function isRecord(x: unknown): x is Record<string, unknown> {
  return x !== null && typeof x === 'object' && !Array.isArray(x);
}

async function validateAndAssertOrder(path: string): Promise<void> {
  const raw = await readFile(path, 'utf8');
  const parsed = JSON.parse(raw) as unknown;
  if (!isRecord(parsed) || typeof parsed.version !== 'string' || !isRecord(parsed.mappings)) {
    throw new Error(
      `Invalid output: expected { version: string, mappings: Record<string, RBiH|RS|HRHB> } at ${path}`
    );
  }
  const mappings = parsed.mappings as Record<string, unknown>;
  if (Object.keys(mappings).length === 0) {
    throw new Error(`Output mapping is empty at ${path}`);
  }
  const keys = Object.keys(mappings);
  const sorted = [...keys].sort((a, b) => a.localeCompare(b));
  if (keys.length !== sorted.length || keys.some((k, i) => k !== sorted[i])) {
    throw new Error(
      `Output mappings are not lexicographically sorted at ${path}. Do not rewrite; fix the generator.`
    );
  }
  for (const [mun_code, controller] of Object.entries(mappings)) {
    if (controller !== null && !CANONICAL_IDS.includes(controller as ControllerId)) {
      throw new Error(
        `Invalid controller for municipality ${mun_code}: ${controller} (must be RBiH, RS, HRHB, or null)`
      );
    }
  }
}

async function main(): Promise<void> {
  const { input } = parseArgs();
  if (!existsSync(input)) {
    throw new Error(
      `Excel not found: ${input}. Expected data/source/1990 to 1995 municipalities_BiH.xlsx or pass --input <path>.`
    );
  }
  if (!existsSync(INDEX_PATH)) {
    throw new Error(
      `Municipality index not found: ${INDEX_PATH}. Run map pipeline to produce municipality_post1995_to_mun1990.json.`
    );
  }

  const mapping = await extractFromExcel(input);
  const sortedEntries = Object.entries(mapping).sort(([a], [b]) => a.localeCompare(b));
  const out = {
    version: 'extract1990_v1',
    mappings: Object.fromEntries(sortedEntries)
  };
  await writeFile(OUTPUT_PATH, JSON.stringify(out, null, 2) + '\n', 'utf8');
  await validateAndAssertOrder(OUTPUT_PATH);
}

main().catch((err) => {
  console.error(err.message || err);
  process.exitCode = 1;
});

File: scripts/data/extract_1990_municipal_winners_from_docx.py
#!/usr/bin/env python3
"""
Phase H1.2.2: Extract 1990 municipal election winners from DOCX → municipality→controller mapping.
Deterministic: no timestamps, stable key ordering, strict party→controller mapping.
Output schema: { "version": "extract1990_v1", "mappings": { "<post1995_code>": "RBiH"|"RS"|"HRHB" } }.
Keys = post1995_code (from municipality_post1995_to_mun1990); values = canonical controller.
Party→controller (from ledger 6B.2): SDA→RBiH, SK-SDP→RBiH, SDS→RS, HDZ BiH→HRHB.
"""

import argparse
import json
import re
import sys
from pathlib import Path

# Optional: python-docx (fail loudly if missing)
try:
    from docx import Document as DocxDocument
except ImportError:
    sys.stderr.write(
        "Error: python-docx is required. Install with: pip install python-docx\n"
    )
    sys.exit(1)

# Canonical controller enum (no null in extraction; unmapped municipalities are omitted)
CONTROLLERS = ("RBiH", "RS", "HRHB")

# Party name / abbreviation → controller (deterministic; do not guess unknown)
PARTY_TO_CONTROLLER = {
    "SDA": "RBiH",
    "SK-SDP": "RBiH",
    "SDS": "RS",
    "HDZ BiH": "HRHB",
    "HDZ": "HRHB",  # common abbreviation
}


def normalize_cell_text(text):
    """Normalize for matching: strip, collapse whitespace, remove footnotes [1], [a], etc."""
    if not text or not isinstance(text, str):
        return ""
    t = text.strip()
    t = re.sub(r"\s+", " ", t)
    t = re.sub(r"\s*\[[^\]]*\]\s*", "", t)  # footnotes [1], [a]
    t = t.strip()
    return t


def normalize_municipality_name(name):
    """Normalize municipality name for lookup: remove entity suffixes (FBiH)/(RS), hyphens→space."""
    t = normalize_cell_text(name)
    t = re.sub(r"\s*\((?:FBiH|RS)\)\s*$", "", t, flags=re.IGNORECASE)
    t = re.sub(r"-", " ", t)
    t = re.sub(r"\s+", " ", t).strip()
    return t


def party_to_controller(party_raw):
    """Map party string to controller. Returns None if unknown (caller should fail)."""
    p = normalize_cell_text(party_raw)
    if not p:
        return None
    # Exact match first
    if p in PARTY_TO_CONTROLLER:
        return PARTY_TO_CONTROLLER[p]
    # Normalize for key match (strip parentheses content for display variants)
    p_norm = re.sub(r"\s*\([^)]*\)\s*", "", p).strip()
    if p_norm in PARTY_TO_CONTROLLER:
        return PARTY_TO_CONTROLLER[p_norm]
    return None


def load_municipality_index(remap_path):
    """
    Load remap JSON and build (normalized_name, display_name) -> [post1995_codes].
    We need to map DOCX municipality names to post1995_code. Remap has post1995_code -> mun1990_name.
    Also rows have post1995_name. Build: for each row, (normalized(mun1990_name), post1995_code), (normalized(post1995_name), post1995_code).
    Return dict: normalized_key -> list of post1995_codes (sorted for determinism).
    """
    path = Path(remap_path)
    if not path.exists():
        raise FileNotFoundError(f"Municipality index/remap not found: {remap_path}")
    data = json.loads(path.read_text(encoding="utf-8"))
    rows = data.get("rows") or []
    index_by_code = data.get("index_by_post1995_code") or {}
    # Build name -> [codes]. From rows we have post1995_code, post1995_name, mun1990_name.
    name_to_codes = {}
    for row in rows:
        code = row.get("post1995_code")
        if not code:
            continue
        for key in (
            row.get("mun1990_name"),
            row.get("post1995_name"),
        ):
            if key:
                norm = normalize_municipality_name(key)
                if norm:
                    name_to_codes.setdefault(norm, []).append(code)
    for code, name in (index_by_code or {}).items():
        if name:
            norm = normalize_municipality_name(name)
            if norm:
                name_to_codes.setdefault(norm, []).append(code)
    # Deduplicate and sort per key
    for k in name_to_codes:
        name_to_codes[k] = sorted(set(name_to_codes[k]))
    return name_to_codes


def get_cell_text(cell):
    """Get cell text (for merged cells, python-docx often gives text only in first cell)."""
    return (cell.text or "").strip()


def find_table_with_headers(doc, municipality_headers=("općina", "opcina"), party_headers=("stranka",)):
    """Find first table that has header row containing municipality and party column names."""
    for table in doc.tables:
        if not table.rows:
            continue
        header_row = table.rows[0]
        header_texts = [get_cell_text(cell) for cell in header_row.cells]
        header_lower = [t.lower() for t in header_texts]
        has_muni = any(
            any(h in t or t in h for h in municipality_headers for t in header_lower)
        )
        has_party = any(
            any(h in t or t in h for h in party_headers for t in header_lower)
        )
        if has_muni and has_party:
            return table, header_texts
    return None, []


def extract_pairs_from_table(table, header_texts):
    """
    Yield (municipality_name, party_name) from table. Skip header row.
    Support 4-column layout: općina left, stranka left, općina right, stranka right.
    Use first two columns as primary; if empty, use next two (or single pair if 2-col).
    """
    municipality_keywords = ("općina", "opcina")
    party_keywords = ("stranka",)
    rows = table.rows
    if not rows:
        return
    for row_idx, row in enumerate(rows):
        if row_idx == 0:
            continue
        cells = row.cells
        if len(cells) < 2:
            continue
        # Try first two columns as (muni, party)
        muni_a, party_a = get_cell_text(cells[0]), get_cell_text(cells[1])
        if muni_a and party_a:
            yield muni_a, party_a
        if len(cells) >= 4:
            muni_b, party_b = get_cell_text(cells[2]), get_cell_text(cells[3])
            if muni_b and party_b:
                yield muni_b, party_b


def main():
    parser = argparse.ArgumentParser(
        description="Extract 1990 municipal winners from DOCX to municipality_political_controllers JSON"
    )
    parser.add_argument("--input", required=True, help="Path to DOCX file")
    parser.add_argument("--output", required=True, help="Path to output JSON file")
    parser.add_argument(
        "--index",
        required=True,
        help="Path to municipality_post1995_to_mun1990.json (for name→code mapping)",
    )
    args = parser.parse_args()

    input_path = Path(args.input)
    if not input_path.exists():
        sys.stderr.write(f"Error: DOCX not found: {input_path}\n")
        sys.exit(1)

    name_to_codes = load_municipality_index(args.index)

    doc = DocxDocument(str(input_path))
    table, header_texts = find_table_with_headers(doc)
    if table is None:
        sys.stderr.write(
            "Error: No table found with Općina/Stranka-style headers in DOCX.\n"
        )
        sys.exit(1)

    # Collect (post1995_code, controller) for each row; allow multiple codes per name (parent expansion)
    mapping = {}
    unknown_parties = set()
    unmatched_municipalities = set()

    for muni_name, party_name in extract_pairs_from_table(table, header_texts):
        controller = party_to_controller(party_name)
        if controller is None:
            if party_name:
                unknown_parties.add(party_name)
            continue
        norm_name = normalize_municipality_name(muni_name)
        if not norm_name:
            continue
        codes = name_to_codes.get(norm_name)
        if not codes:
            unmatched_municipalities.add(muni_name)
            continue
        for code in codes:
            mapping[code] = controller

    if unknown_parties:
        sys.stderr.write(
            f"Error: Unknown party names (cannot map to RBiH/RS/HRHB): {sorted(unknown_parties)}\n"
        )
        sys.exit(1)

    if not mapping:
        sys.stderr.write(
            "Error: Mapping is empty. Check DOCX table structure and index.\n"
        )
        sys.exit(1)

    # Deterministic output: sort keys
    out = {
        "version": "extract1990_v1",
        "mappings": dict(sorted(mapping.items())),
    }
    output_path = Path(args.output)
    output_path.parent.mkdir(parents=True, exist_ok=True)
    output_path.write_text(
        json.dumps(out, indent=2, ensure_ascii=False) + "\n",
        encoding="utf-8",
    )
    if unmatched_municipalities:
        sys.stderr.write(
            f"Warning: Unmatched municipalities (not in index): {sorted(unmatched_municipalities)}\n"
        )


if __name__ == "__main__":
    main()
